x-pack/plugins/apm/server/lib/search_strategies/correlations/query_correlation.ts

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { estypes } from '@elastic/elasticsearch';
+
+import type { ElasticsearchClient } from 'src/core/server';
+
+import { TRANSACTION_DURATION } from '../../../../common/elasticsearch_fieldnames';
+import type { SearchServiceParams } from '../../../../common/search_strategies/correlations/types';
+
+import { getQueryWithParams } from './get_query_with_params';
+import { PERCENTILES_STEP } from './constants';
+
+export interface HistogramItem {
+  key: number;
+  doc_count: number;
+}
+
+interface ResponseHitSource {
+  [s: string]: unknown;
+}
+interface ResponseHit {
+  _source: ResponseHitSource;
+}
+
+interface BucketCorrelation {
+  buckets_path: string;
+  function: {
+    count_correlation: {
+      indicator: {
+        expectations: number[];
+        doc_count: number;
+      };
+    };
+  };
+}
+
+export const getTransactionDurationCorrelationRequest = (
+  params: SearchServiceParams,
+  percentiles: number[],
+  totalHits: number,
+  fieldName?: string,
+  fieldValue?: string
+): estypes.SearchRequest => {
+  const query = getQueryWithParams({ params, fieldName, fieldValue });
+
+  const ranges = percentiles.reduce(
+    (p, to) => {
+      const from = p[p.length - 1].to;
+      p.push({ from, to });
+      return p;
+    },
+    [{ to: 0 }] as Array<{ from?: number; to?: number }>
+  );
+  ranges.push({ from: ranges[ranges.length - 1].to });
+
+  const step = PERCENTILES_STEP;
+  const tempPercentiles = [percentiles[0]];
+  const tempFractions = [step / 100];
```

Comment:
Calculating fractions like this will not work. We have discovered in certain datasets/scenarios we end up with pretty bad fractions since the percentiles can be exceptionally approximate. 

We need to use fractions for two things:

- 1. calculate the expectations for the bucketed pearson correlation. For that use, using these deduplicated fractions (based on percentile step) is good.
- 2. Pass to the k-s test as a representation of the distribution of the data. For this, it's no good. We need to provide the "true" fractions. To do that we need to grab the ranges (your `tempRanges` when corrected to use the `tempPercentiles`) execute a search against `TRANSACTION_DURATION` (using whatever filters there are in the context). This gives you the true count for each percentile range, then you can sum them up, and divide each individual range bucket count by the total to get your "true fraction" for the KS test.

Reply:
As discussed in Slack, I will address this in the next follow up.